src/main/scala/io/iohk/ethereum/consensus/blocks/BlockGenerator.scala

```diff
@@ -0,0 +1,62 @@
+package io.iohk.ethereum.consensus.blocks
+
+import akka.util.ByteString
+import io.iohk.ethereum.domain.{Address, Block, SignedTransaction}
+import io.iohk.ethereum.ledger.BlockPreparationError
+
+/**
+ * We use a `BlockGenerator` to create the next block.
+ * In a PoW setting, this is what a miner typically does.
+ * In general, a [[BlockGenerator]] depends on and is provided by the
+ * [[io.iohk.ethereum.consensus.Consensus Consensus]].
+ *
+ * @note This is generally a stateful object.
+ *
+ * @see [[io.iohk.ethereum.consensus.Consensus#blockGenerator Consensus#blockGenerator()]],
+ *      [[io.iohk.ethereum.ledger.BlockPreparator BlockPreparator]]
+ *
+ */
+trait BlockGenerator {
+  /**
+   * The type of consensus-specific data used in the block generation process.
+   * For example, under [[io.iohk.ethereum.consensus.ethash.EthashConsensus EthashConsensus]],
+   * this represents the [[io.iohk.ethereum.network.p2p.messages.PV62.BlockBody#uncleNodesList ommers]].
+   */
+  type X
```

Comment:
I don't quite like this idea. Can't we do something like this instead:
`def generateBlock[Data <: BlockGenerationData](..., data: Data)`
and for ethash, for example: `case class EthashBlockGenerationData(ommers: Ommers): extends BlockGenerationData` or `RaftBlockGenerationData(raft specific stuff for generating blocks)`

Reply:
Why don't you like it? 
We naturally express the idea that the shape of "ommers" (or whatever concept applies) depends on the `BlockGenerator`, so its type must be a property of it.